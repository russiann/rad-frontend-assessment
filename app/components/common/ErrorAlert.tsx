import React from "react";
import { Button } from "@heroui/react";
import { Icon } from "@iconify/react";

type ErrorAlertProps = {
  error: string;
  onDismiss?: () => void;
  variant?: 'default' | 'inline';
  className?: string;
};

export function ErrorAlert({ 
  error, 
  onDismiss, 
  variant = 'default',
  className = '' 
}: ErrorAlertProps) {
  const baseClasses = "p-3 bg-danger-50 border border-danger-200 rounded-lg";
  const finalClasses = `${baseClasses} ${className}`;

  if (variant === 'inline') {
    return (
      <div className={finalClasses}>
        <div className="flex items-center gap-2">
          <Icon icon="lucide:alert-circle" className="text-danger-500 flex-shrink-0" />
          <span className="text-danger-700 text-sm">{error}</span>
        </div>
      </div>
    );
  }

  return (
    <div className={finalClasses}>
      <div className="flex items-center justify-between">
        <div className="flex items-center gap-2">
          <Icon icon="lucide:alert-circle" className="text-danger-500 flex-shrink-0" />
          <span className="text-danger-700 text-sm">{error}</span>
        </div>
        {onDismiss && (
          <Button
            isIconOnly
            size="sm"
            variant="light"
            color="danger"
            onPress={onDismiss}
          >
            <Icon icon="lucide:x" />
          </Button>
        )}
      </div>
    </div>
  );
} 