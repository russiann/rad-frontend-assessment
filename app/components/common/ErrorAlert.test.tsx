import { describe, it, expect, vi } from 'vitest'
import { render, screen } from '@testing-library/react'
import { userEvent } from '@testing-library/user-event'
import { ErrorAlert } from './ErrorAlert'

describe('ErrorAlert', () => {
  describe('snapshots', () => {
    it('renders default variant correctly', () => {
      const { container } = render(
        <ErrorAlert error="Default error" />
      )
      expect(container).toMatchSnapshot()
    })

    it('renders inline variant correctly', () => {
      const { container } = render(
        <ErrorAlert error="Inline error" variant="inline" />
      )
      expect(container).toMatchSnapshot()
    })

    it('renders with dismiss button correctly', () => {
      const mockDismiss = vi.fn()
      const { container } = render(
        <ErrorAlert 
          error="Error with dismiss" 
          onDismiss={mockDismiss} 
        />
      )
      expect(container).toMatchSnapshot()
    })

    it('renders with custom className correctly', () => {
      const { container } = render(
        <ErrorAlert 
          error="Error with custom class" 
          className="custom-class" 
        />
      )
      expect(container).toMatchSnapshot()
    })
  })

  describe('rendering', () => {
    it('displays error message', () => {
      render(<ErrorAlert error="Test error message" />)
      expect(screen.getByText('Test error message')).toBeInTheDocument()
    })

    it('displays alert icon', () => {
      const { container } = render(<ErrorAlert error="Test error" />)
      const iconElement = container.querySelector('span:first-child')
      expect(iconElement).toBeInTheDocument()
    })

    it('applies default variant styling by default', () => {
      render(<ErrorAlert error="Test error" />)
      const container = screen.getByText('Test error').closest('div')?.parentElement?.parentElement
      expect(container).toHaveClass('p-3', 'bg-danger-50', 'border', 'border-danger-200', 'rounded-lg')
    })

    it('applies custom className when provided', () => {
      render(<ErrorAlert error="Test error" className="my-custom-class" />)
      const container = screen.getByText('Test error').closest('div')?.parentElement?.parentElement
      expect(container).toHaveClass('my-custom-class')
    })
  })

  describe('inline variant', () => {
    it('renders inline variant without dismiss button', () => {
      render(<ErrorAlert error="Inline error" variant="inline" />)
      expect(screen.queryByRole('button')).not.toBeInTheDocument()
    })

    it('has simplified layout for inline variant', () => {
      render(<ErrorAlert error="Inline error" variant="inline" />)
      const errorContainer = screen.getByText('Inline error').parentElement
      expect(errorContainer).toHaveClass('flex', 'items-center', 'gap-2')
    })
  })

  describe('default variant', () => {
    it('renders without dismiss button when onDismiss not provided', () => {
      render(<ErrorAlert error="Default error" />)
      expect(screen.queryByRole('button')).not.toBeInTheDocument()
    })

    it('renders with dismiss button when onDismiss provided', () => {
      const mockDismiss = vi.fn()
      render(<ErrorAlert error="Default error" onDismiss={mockDismiss} />)
      expect(screen.getByRole('button')).toBeInTheDocument()
    })

    it('has justify-between layout for default variant', () => {
      render(<ErrorAlert error="Default error" />)
      const errorContainer = screen.getByText('Default error').parentElement?.parentElement
      expect(errorContainer).toHaveClass('flex', 'items-center', 'justify-between')
    })
  })

  describe('dismiss functionality', () => {
    it('calls onDismiss when dismiss button is clicked', async () => {
      const user = userEvent.setup()
      const mockDismiss = vi.fn()
      
      render(<ErrorAlert error="Dismissible error" onDismiss={mockDismiss} />)
      
      const dismissButton = screen.getByRole('button')
      await user.click(dismissButton)
      
      expect(mockDismiss).toHaveBeenCalledOnce()
    })

    it('does not render dismiss button in inline variant even with onDismiss', () => {
      const mockDismiss = vi.fn()
      render(
        <ErrorAlert 
          error="Inline error" 
          variant="inline" 
          onDismiss={mockDismiss} 
        />
      )
      expect(screen.queryByRole('button')).not.toBeInTheDocument()
    })
  })

  describe('accessibility', () => {
    it('dismiss button has proper attributes', () => {
      const mockDismiss = vi.fn()
      render(<ErrorAlert error="Error with dismiss" onDismiss={mockDismiss} />)
      
      const button = screen.getByRole('button')
      expect(button).toHaveAttribute('type', 'button')
    })

    it('error text has proper styling for readability', () => {
      render(<ErrorAlert error="Accessible error" />)
      const errorText = screen.getByText('Accessible error')
      expect(errorText).toHaveClass('text-danger-700', 'text-sm')
    })
  })
})