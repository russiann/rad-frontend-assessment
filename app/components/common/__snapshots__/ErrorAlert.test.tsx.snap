// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ErrorAlert > snapshots > renders default variant correctly 1`] = `
<div>
  <div
    class="p-3 bg-danger-50 border border-danger-200 rounded-lg "
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center gap-2"
      >
        <span />
        <span
          class="text-danger-700 text-sm"
        >
          Default error
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`ErrorAlert > snapshots > renders inline variant correctly 1`] = `
<div>
  <div
    class="p-3 bg-danger-50 border border-danger-200 rounded-lg "
  >
    <div
      class="flex items-center gap-2"
    >
      <span />
      <span
        class="text-danger-700 text-sm"
      >
        Inline error
      </span>
    </div>
  </div>
</div>
`;

exports[`ErrorAlert > snapshots > renders with custom className correctly 1`] = `
<div>
  <div
    class="p-3 bg-danger-50 border border-danger-200 rounded-lg custom-class"
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center gap-2"
      >
        <span />
        <span
          class="text-danger-700 text-sm"
        >
          Error with custom class
        </span>
      </div>
    </div>
  </div>
</div>
`;

exports[`ErrorAlert > snapshots > renders with dismiss button correctly 1`] = `
<div>
  <div
    class="p-3 bg-danger-50 border border-danger-200 rounded-lg "
  >
    <div
      class="flex items-center justify-between"
    >
      <div
        class="flex items-center gap-2"
      >
        <span />
        <span
          class="text-danger-700 text-sm"
        >
          Error with dismiss
        </span>
      </div>
      <button
        class="z-0 group relative inline-flex items-center justify-center box-border appearance-none select-none whitespace-nowrap font-normal subpixel-antialiased overflow-hidden tap-highlight-transparent transform-gpu data-[pressed=true]:scale-[0.97] cursor-pointer outline-hidden data-[focus-visible=true]:z-10 data-[focus-visible=true]:outline-2 data-[focus-visible=true]:outline-focus data-[focus-visible=true]:outline-offset-2 text-tiny gap-2 rounded-small px-0 !gap-0 transition-transform-colors-opacity motion-reduce:transition-none bg-transparent text-danger data-[hover=true]:bg-danger/20 min-w-8 w-8 h-8"
        data-react-aria-pressable="true"
        tabindex="0"
        type="button"
      >
        <span />
      </button>
    </div>
  </div>
</div>
`;
